initial readme for grpc project (#51)
CodeYourFuture/immersive-go-course: grpc-client-server/README.md

```console
> ls prober
> prober.pb.go		prober.proto		prober_grpc.pb.go
```

Read through `prober_grpc.pb.go` - this is the interface we will use in our code. This is how gRPC works: a `proto` format 
gets generated into language-specific code that we can use to interact with gRPCs. If we're working with multiple programming languages
that need to interact through RPCs, we can do this by generating from the same protocol buffer definition with the language-specific
tooling. 

Now run the server and client code. You should see output like this.

```console
> go run prober_server/main.go
> 2022/10/19 17:51:32 server listening at [::]:50051
````

```console
> go run prober_client/main.go 
> 2022/10/19 17:52:15 Response Time: 117.000000
```

We've now gained some experience with the protocol buffer format, learned 
how to generate Go code from protocol buffer definitions, and call that code from Go programs.

## Implement prober logic

Let's modify the prober service slightly. Instead of the simple one-off HTTP GET against a hardcoded google.com, we are going to modify the service to probe a HTTP endpoint N times and return the average time to GET that endpoint to the client. 

> Change your prober request and response to look like this and regenerate your generated go code.
> ```
> // The request message
> message ProbeRequest {
>   string endpoint = 1;    // Endpoint to probe
>   int32 repetitions = 2;  // How many probes to execute
> }
> 
> // The response message 
> message ProbeReply {
>   float result = 1;    // Average response time (in milliseconds)
> }
> ```

Reviewer: I'd be tempted to just instruct rather than show here:

```suggestion
Change your prober request and response:
* Add a field to the `ProbeRequest` for the number of requests to make.
* Rename the field in `ProbeReply` (and perhaps add a comment) to make clear it's the _average_ response time of the several requests.

Note that it's ok to rename fields in protobuf (unlike when we use JSON), because the binary encoding of protobuf messages doesn't include field names. You can [read more about backward/forward compatibility with protobufs](https://earthly.dev/blog/backward-and-forward-compatibility/) if you want.

Remember that you'll need to re-generate your Go code after changing your proto definitions.
```

The initial version of the code demonstrates how to use the standard [`net/http` package](https://pkg.go.dev/net/http) and the standard time package.

Add up all the elapsed times, divide by the number of repetitions, and return the average to the client.
The client should print out the average value received.
You can do arithmetic operations like addition and division on `time.Duration` values in Go.

## Add a client timeout

Maybe the site we are probing is very slow (which can happen for all kinds of reasons, from network problems to excessive load), 
or perhaps the number of repetitions is very high.
Either way, we never want our program to wait forever. 
If we are not careful about preventing this then we can end up building systems where problems in one small part of the system 
spread across all of the services that talk to that part of the system. 

On the client side, add a [timeout](https://pkg.go.dev/context#WithTimeout) to stop waiting after 1 second.

Run your client against some website - how many repetitions do you need to see your client timeout?

## Handling Errors

How do we know if the HTTP fetch succeeded at the server? Add a check to make sure it did.

> How should we deal with errors, i.e. if the endpoint isn't found?

Reviewer: ```suggestion
How should we deal with errors, e.g. if the endpoint isn't found, or says the server is in an error state?
```

Modify your code and proto format to handle these cases.

## Extra Challenge: Serve and Collect Prometheus Metrics

These sections are optional - do them for an extra challenge if time permits.

### Part 1: Add Prometheus Metrics
Let's learn something about how to monitor applications.

In software operations, we want to know what our software is doing and how it is performing.
One very useful technique is to have our program export metrics. Metrics are basically values that your 
program makes available (the industry standard is to export and scrape over HTTP). 

Specialised programs, such as Prometheus, can then fetch metrics regularly

> from all the running instances of yur program, store the history of these metrics, and do useful arithmetic on them

Reviewer: ```suggestion
from all the running instances of your program, store the history of these metrics, and do useful arithmetic on them
```

# GRPC Client and Server Communication 

The goals of this project are: 

>  * to understand the differences between gRPC and REST

Author: and motivations

Read the [gRPC Introduction](https://grpc.io/docs/what-is-grpc/introduction/) and 
[gRPC Core Concepts](https://grpc.io/docs/what-is-grpc/core-concepts/) for an overview of gRPC.
todo stuff from primer

## Get started with a HelloWorld gRPC example

Start by building a simple gRPC ‘hello world’ server and cli client in golang. See [gRPC Quickstart](https://grpc.io/docs/languages/go/quickstart/) for detailed instructions.

> ## Change HelloWorld to Prober

Author: so i can always lay out a sample project to get them here but I quite like teaching the implicit skill of find an example chunk of code and hack on it... your call, tom

Reviewer 2: I think the "don't do this in real life" instruction further makes me feel less good about this. Is there a way to teach it that _doesn't_ require renaming?

Then update your server to execute the probe: do a HTTP fetch of `endpoint` the specified number of times.
Use the standard [`net/http` package](https://pkg.go.dev/net/http).

Time each fetch using the `time` package:

```
  	start := time.Now()
	resp, _:= http.Get(endpoint)
    elapsed := time.Since(start)
```

Add up all the elapsed times, divide by the number of repetitions, and return the average to the client.
The client should print out the average value received.
You can do arithmetic operations like addition and division on `time.Duration` values in Go.
However, under the hood `time.Duration` is an `int64`. To divide by an `int32` and return a `float` will need to do some [type conversions](https://go.dev/tour/basics/13).

## Add a client timeout

Maybe the site you are probing is very slow, or the number of repetitions is very high.
Either way, you don't want your program to wait forever. 

On the client side, add a [timeout](https://pkg.go.dev/context#WithTimeout) to stop waiting after 1 second].

Run your client against some website - how many repetitions do you need to see your client timeout?

> ## Extra Challenges: Errors, Continuous Probing, Serve Prometheus Metrics

Author: I've been less specific here about instructions, to give them a chance to swim by themselves. 
Again, would like feedback about whether that's the right choice.

Reviewer 2: I think that this project is short enough that some of this should be in the project, not an extension.

# GRPC Client and Server Communication 

The goals of this project are: 
 * to use gRPC requests, and understand how they differ from REST
 * to use deadlines with gRPCs
 * to build a simple prober
 * optionally, gain experience with instrumenting an application and running Prometheus

## What is gRPC?

> RPCs aren’t the same as HTTP requests. HTTP is a text-based protocol, oriented towards resources or entities. RPCs are oriented towards calling a specific function. RPCs are highly structured; normally the request and response are encoded in efficient binary formats which are not human-readable. HTTP APIs are usually limited to CRUD (Create, Read, Update, Delete) operations and RPCs can perform any kind of operation. If you want to efficiently integrate two systems that you control, RPCs are a good choice. If you want to provide an API for use by developers outside of your organisation then HTTP APIs are generally a better choice, because they are simpler to develop against and all programming languages provide good HTTP support. gRPC, which we will use in this exercise, is a RPC implementation that is fairly common in industry. 

Reviewer 2: I think this paragraph is confusing if the distinction between (for example) REST, HTTP, protocols, architectural patterns and specific implementations isn't clear — which I think it won't be for the trainees. They may not be aware they've been doing diet-REST and certainly we haven't introduced it. I suggest that, if we want to compare to REST, we need to give an introduction. Otherwise, I would just say something in the direction of:

* We've used simple HTTP request so far, making and handling `GET` and `POST` requests
* RPC is another way to communicate between clients and servers
* It is used commonly *within* an organisation, where it has advantages around types, performance, and streaming
* gRPC is not used to communicate from web browsers to servers (important because this is context they have)

I think your other points are good, around resource vs procedure and external APIs.

Read the [gRPC Introduction](https://grpc.io/docs/what-is-grpc/introduction/) and 
[gRPC Core Concepts](https://grpc.io/docs/what-is-grpc/core-concepts/) for an overview of gRPC.
todo stuff from primer

## Get started with a HelloWorld gRPC example

Start by building a simple gRPC ‘hello world’ server and cli client in golang. See [gRPC Quickstart](https://grpc.io/docs/languages/go/quickstart/) for detailed instructions.

## Change HelloWorld to Prober

> Once you have the 'hello world' example from the gRPC Quickstart tutorial working, we will move on to implementing a prober.

Reviewer 2: The concepts of a prober needs introducing

Rename `greeter_client` and `greeter_server` to `prober_client` and `prober_server`.   
Change the `helloworld` subdirectory to be called `prober`,` and helloworld.proto` to `prober.proto`.

> Delete the genereted .go files in the same directory as `prober.proto`.

Reviewer 2: This will compress into one line. Better as bullets?

Now update the the prober.proto file so that the main content section looks like this:

```
package prober;

// The greeting service definition.
service Prober {
  // Sends a greeting
  rpc DoProbes (ProbeRequest) returns (ProbeReply) {}
}

// The request message containing the user's name.
message ProbeRequest {
  string name = 1;
}

// The response message containing the greetings
message ProbeReply {
  string message = 1;
}
```

Regenerate the generated protobuf code: 
```
protoc --go_out=. --go_opt=paths=source_relative \
    --go-grpc_out=. --go-grpc_opt=paths=source_relative prober/prober.proto

> ```

Reviewer 2: To link the ideas from the introduction and hello world to this, I think it's worth reiterating at each stage: _what_ they are doing and _why_ it's important at this stage.

Observe the new generated files:
```

> ls prober

Reviewer 2: `> ls proper` to match style elsewhere, same applies to other commands.

Using ` ```console ` adds a bit of highlighting.

prober.pb.go		prober.proto		prober_grpc.pb.go
```

> Read through `prober_grpc.pb.go` - this is the interface you will use in your code. This is how gRPC works: a `proto` format 

Reviewer 2: Stylistic but `you` and `your` → to `we` and `our` please, unless giving direct instructions (e.g. `You need to set up account`, `You should not...`). I can see you've used "we" elsewhere.

Note to self: need to add this to `CONTRIBUTING.md`.

gets generated into language-specific code that you can use to interact with gRPCs. If you're working with multiple programming languages
that need to interact through RPCs, you can do this by generating from the same protocol buffer definition with the language-specific
tooling. 

Now, update the client and server `main.go` files to use the changed package, service and message names.
Get these running again as they did originally. The functionality will still be the same for now, just the hello world interaction.

```
go run prober_server/main.go
2022/10/09 20:20:55 server listening at [::]:50051
2022/10/09 20:21:24 Received: world
````

```
go run prober_client/main.go 
2022/10/09 20:21:24 Greeting: Hello world
```

While we haven't yet changed what the code does, we've gained some experience with the protocol buffer format, learned 
how to generate Go code from protocol buffer definitions, and call that code from Go programs.

> Note: in a real production system you can't rename protocol buffers as we've done here: you would run into problems because your old clients can't talk to your new servers, and vice-versa. Since we gradually release code, we would have to support both services until all old clients were turned down.

Reviewer 2: This feels like a bit of an educational anti-pattern: to get them to do something and they say "don't actually do this". Is there a way to structure this project to avoid it?

## Implement prober logic

Let's modify the prober service slightly. Instead of the simple HelloWorld greeting, we are going to implement a service that will probe a HTTP endpoint N times and return the average time to fetch that endpoint to the client. 

This tool could be useful in a real operational context. Imagine that we want to verify that our site is available and has acceptable latency from many different locations around the world. We might run several instances of our prober server in
different regions. 

Change your prober request and response to look like this and regenerate your generated go code.

> ```

Reviewer 2: ` ```protobuf `

// The request message
message ProbeRequest {
  string endpoint = 1;    // Endpoint to probe
  int32 repetitions = 2;  // How many probes to execute
}

// The response message 
message ProbeReply {
  float average = 1;    // Average response time (in milliseconds)
}
```

Update your client to read the endpoint and number of repetitions from the [command line](https://gobyexample.com/command-line-arguments).

> Then update your server to execute the probe: do a HTTP fetch of `endpoint` the specified number of times.

Reviewer 2: fetch → `GET` IMO

"do a HTTP fetch of `endpoint` the specified number of times" is also quite colloquial — possibly not clear to a non-native speaker.

Use the standard [`net/http` package](https://pkg.go.dev/net/http).

Time each fetch using the `time` package:

```
  	start := time.Now()

> 	resp, _:= http.Get(endpoint)

Reviewer 2: To an extent, this gives the answer for the instructions above. It also drops error handling; I think we shouldn't teach that.

    elapsed := time.Since(start)
```

Add up all the elapsed times, divide by the number of repetitions, and return the average to the client.
The client should print out the average value received.
You can do arithmetic operations like addition and division on `time.Duration` values in Go.

> However, under the hood `time.Duration` is an `int64`. To divide by an `int32` and return a `float` will need to do some [type conversions](https://go.dev/tour/basics/13).

Reviewer 2: This last line is an exercise for the reader to discover, I think.

## Add a client timeout

Maybe the site you are probing is very slow, or the number of repetitions is very high.

> Either way, you don't want your program to wait forever. 

Reviewer 2: Why not? When does this happen? Does this generalise?

On the client side, add a [timeout](https://pkg.go.dev/context#WithTimeout) to stop waiting after 1 second].

Run your client against some website - how many repetitions do you need to see your client timeout?

## Extra Challenges: Errors, Continuous Probing, Serve Prometheus Metrics

These sections are optional - do them for an extra challenge if time permits.

### Part 1:
How do we know if the HTTP fetch succeeded at the server? Add a check to make sure it did.
How should we deal with errors, i.e. if the endpoint isn't found?

> Modify your code and proto format to handle these cases.

Reviewer 2: In my view, error handling shouldn't be an extension. In other exercises, tests are left unaddressed which I think is _just about_ OK (and purely a time-available-for-me thing), but error handling not so much.

### Part 2: Run probes on a schedule
Instead of getting probe times via the gRPC client we might consider running the endpoint probes on a schedule. Change the
prober_server to accept a frequency argument instead of the old `repetitions` parameter, 

> and continue running probes every N seconds forever. 

Reviewer 2: Why would we do this? Any good resources?

Change vs add an option? Goes back to managing change through time, and avoiding breaking changes. Can we teach the good habit? We don't have to be explicit about why, tho that would be nice.

You will need to use a [goroutine](https://gobyexample.com/goroutines) to keep running probes in the background.

Now, add a new client operation to get the value of the last probe run.
To make this work, you will need to pass a value to your prober goroutine that it can use to store the most recent latency.
You can store this value in your `server` struct.
You will also need to use a `sync.Mutex` to guarantee that the goroutine that serves your client request sees the last value 

> written by the prober goroutine. A Tour of Go has a good example of [using a sync.Mutex](https://go.dev/tour/concurrency/9).

Reviewer 2: First time specifically mentioning a goroutine and `Mutex`. This absolutely needs explanation or an external link (possibly a whole project) — I don't think the tour of Go does a good job of it.

### Part 3: Add Prometheus Metrics

> If you still have time, then let's learn something about how to monitor applications.

Reviewer 2: Minor: I think we don't need to worry about their time; we can just lay out the exercise and let them decide whether to take it on.

In software operations, we want to know what our software is doing and how it is performing.
One very useful technique is to have your program export metrics. Metrics are basically values that your 

> program makes available (via HTTP). Specialised programs, such as Prometheus, can then fetch metrics regularly

Reviewer 2: This draws a metrics-HTTP link, as if metrics _have to be_ exposed over HTTP. Is that correct, or a implementation detail here? Perhaps `(via HTTP)` → `to external systems such as...`

from all the running instances of your program, store the history of these metrics, and do useful arithmetic on them
(like computing rates, averages, and maximums). You can use this data to do troubleshooting and to alert if things 
go wrong.

Read the [Overview of Prometheus](https://prometheus.io/docs/introduction/overview/).

Now add Prometheus metrics to your prober server. Every time you execute a probe, update a `gauge` metric that tracks the latency.
Add a `label` specifying the endpoint being probed. 
The [Prometheus Guide to Instrumenting a Go Application](https://prometheus.io/docs/guides/go-application/) has all the information you need to do this.

Once you've run your program, use your client to execute periodic probes against some endpoint. 
Now use the `curl` program or your browser to view the metrics on the port that you've chosen to serve them on.
You should see a number of built-in Go metrics, plus your new gauge.

If you use your client to start probing a second endpoint, you should see a second labelled metric appear.

> ### Part 3: Scrape Prometheus Metrics

Reviewer 2: Part 4

(We haven't really used the concept of "parts" so far. Not averse but I feel like part 1 is non negotiable, 2 is standalone, and 3+4 are a logical pair)

The final step is to set up the Prometheus application to periodically pull metrics from your `prober_server`.

You'll need to set up a simple configuration to scrape your `prober_server`. Prometheus [configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/) is quite complex but you can adapt this [example configuration](https://github.com/prometheus/prometheus/blob/main/documentation/examples/prometheus.yml).

Run Prometheus in Docker locally - see [Prometheus Installation](https://prometheus.io/docs/prometheus/latest/installation/).

> Use a shared volume for the configuration file, as suuggested in the guide above.

Reviewer 2: The projects thus far have either run entirely in docker, or not in docker at all. I think it's worth introducing here that it's possible (and can be easier?) to run systems like this in docker, rather than install them on your machine.
